File: api/_dedupe_logs.py
"""_dedupe_logs.py — build duplicate/related groupings from relatedTo field.

Deterministic. No LLM call. Part of Layer 3 tools.
Reads the LLM's relatedTo output and builds bidirectional groupings.
"""

from collections import defaultdict


def dedupe_logs(results: list[dict]) -> list[dict]:
    """Ensure relatedTo is bidirectional — if A lists B, B lists A.

    Also builds group IDs so the UI can cluster related entries together.

    Args:
        results: List of triage result dicts with 'id' and 'relatedTo' fields.

    Returns:
        Same list with bidirectional relatedTo and a 'groupId' field added.
    """
    # Build adjacency: for each result, ensure bidirectional links
    by_id = {r["id"]: r for r in results}
    id_set = set(by_id.keys())

    for result in results:
        related = result.get("relatedTo", [])
        # Filter out invalid IDs references
        related = [rid for rid in related if rid in id_set and rid != result["id"]]
        result["relatedTo"] = sorted(set(related))

    # Make bidirectional
    for result in results:
        for related_id in result["relatedTo"]:
            other = by_id.get(related_id)
            if other and result["id"] not in other.get("relatedTo", []):
                other["relatedTo"].append(result["id"])
                other["relatedTo"] = sorted(set(other["relatedTo"]))

    # Assign group IDs via connected components (Union-Find)
    parent = {r_id: r_id for r_id in id_set}

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(x, y):
        px, py = find(x), find(y)
        if px != py:
            parent[px] = py

    for result in results:
        for related_id in result["relatedTo"]:
            union(result["id"], related_id)

    groups = defaultdict(list)
    for result in results:
        root = find(result["id"])
        groups[root].append(result["id"])

    for result in results:
        root = find(result["id"])
        result["groupId"] = f"group-{root}"

    return results

File: lib/dedupeLogs.ts
import type { TriageResult } from "./types";

export function dedupeLogs(results: TriageResult[]): TriageResult[] {
  const byId = new Map<string, TriageResult>();
  for (const r of results) byId.set(r.id, r);
  const idSet = byId;

  for (const result of results) {
    let related = (result.relatedTo || []).filter(
      (rid) => idSet.has(rid) && rid !== result.id
    );
    result.relatedTo = [...new Set(related)].sort();
  }

  for (const result of results) {
    for (const relatedId of result.relatedTo) {
      const other = byId.get(relatedId);
      if (other && !other.relatedTo.includes(result.id)) {
        other.relatedTo.push(result.id);
        other.relatedTo = [...new Set(other.relatedTo)].sort();
      }
    }
  }

  const parent = new Map<string, string>();
  for (const r of results) parent.set(r.id, r.id);

  function find(x: string): string {
    let root = x;
    while (parent.get(root) !== root) {
      root = parent.get(root)!;
    }
    let curr = x;
    while (parent.get(curr) !== root) {
      const next = parent.get(curr)!;
      parent.set(curr, root);
      curr = next;
    }
    return root;
  }

  function union(x: string, y: string): void {
    const px = find(x);
    const py = find(y);
    if (px !== py) parent.set(px, py);
  }

  for (const result of results) {
    for (const relatedId of result.relatedTo) {
      union(result.id, relatedId);
    }
  }

  for (const result of results) {
    result.groupId = `group-${find(result.id)}`;
  }

  return results;
}
